File: RESTFul/CMS/Java/app1/src/main/java/com/att/api/cms/service/CMSService.java
/* vim: set expandtab tabstop=4 shiftwidth=4 softtabstop=4 foldmethod=marker */

/*
 * ====================================================================
 * LICENSE: Licensed by AT&T under the 'Software Development Kit Tools
 * Agreement.' 2013.
 * TERMS AND CONDITIONS FOR USE, REPRODUCTION, AND DISTRIBUTIONS:
 * http://developer.att.com/sdk_agreement/
 *
 * Copyright 2013 AT&T Intellectual Property. All rights reserved.
 * For more information contact developer.support@att.com
 * ====================================================================
 */

package com.att.api.cms.service;

import java.text.ParseException;
import java.util.Map;

import org.json.JSONObject;

import com.att.api.oauth.OAuthToken;
import com.att.api.rest.APIResponse;
import com.att.api.rest.RESTClient;
import com.att.api.rest.RESTException;
import com.att.api.service.APIService;

/**
 * Used to interact with version 1 of the Call Management API.
 *
 * @author <a href="mailto:pk9069@att.com">Pavel Kazakov</a>
 * @version 3.0
 * @since 2.2
 * @see <a href="https://developer.att.com/docs/apis/rest/1/Call%20Management%20(Beta)">CMS Documentation</a>
 */
public class CMSService extends APIService {

    /**
     * Creates a CMSService object.
     *
     * @param fqdn fully qualified domain name to use for sending requests
     * @param token OAuth token to use for authorization
     */
    public CMSService(String fqdn, OAuthToken token) {
        super(fqdn, token);
    }

    /**
     * Internal method used to convert a map of variables to a JSON string.
     *
     * @param vars map to convert
     * @return JSON string
     */
    private String buildScriptVariables(final Map<String, String> vars) {
        final JSONObject scriptVars = new JSONObject();
        for (final Map.Entry<String, String> entry : vars.entrySet()) {
            scriptVars.put(entry.getKey(), entry.getValue());
        }
        return scriptVars.toString();
    }

    /**
     * Sends a request to the API for creating a CMS session.
     *
     * @param variables variables to send
     * @return CMSSessionResponse API response
     * @throws RESTException if API request was not successful
     */
    public CMSSessionResponse createSession(Map<String, String> variables)
            throws RESTException {

        String endpoint = getFQDN() + "/rest/1/Sessions";
        String body = buildScriptVariables(variables);
        APIResponse response = new RESTClient(endpoint)
            .addHeader("Content-Type", "application/json")
            .addHeader("Accept", "application/json")
            .addAuthorizationHeader(getToken())
            .httpPost(body);

        try {
            String responseBody = response.getResponseBody();
            JSONObject jsonResponse = new JSONObject(responseBody);
            final String id = jsonResponse.getString("id");
            final boolean success = jsonResponse.getBoolean("success");
            return new CMSSessionResponse(id, success);
        } catch (ParseException pe) {
            throw new RESTException(pe);
        }
    }

    /**
     * Sends a request to the API for sending a signal to an existing session.
     *
     * <p>
     * A session may be created with <code>createSession()</code>.
     * </p>
     *
     * @param sessionId session id to use for sending signal
     * @param signal signal to send
     * @return status
     * @throws RESTException if API request was not successful
     * @see #createSession(java.util.Map)
     */
    public String sendSignal(String sessionId, String signal)
            throws RESTException {

        if (sessionId == null || sessionId.equals("")) {
            String err = "Please create a session before sending a signal";
            throw new RESTException(err);
        }

        String fqdn = getFQDN();
        String endpoint = fqdn + "/rest/1/Sessions/" + sessionId + "/Signals";
        JSONObject msgBody = new JSONObject();
        msgBody.put("signal", signal);

        APIResponse apiResponse = new RESTClient(endpoint)
            .setHeader("Accept", "application/json")
            .setHeader("Content-Type", "application/json")
            .addAuthorizationHeader(getToken())
            .httpPost(msgBody.toString());

        final String responseBody = apiResponse.getResponseBody();
        try {
            final JSONObject jsonResponse = new JSONObject(responseBody);
            final String status = jsonResponse.getString("status");
            return status;
        } catch (ParseException pe) {
            throw new RESTException(pe);
        }
    }
}

File: RESTFul/CMS/PHP/app1/lib/CMS/CMSService.php
<?php
namespace Att\Api\CMS;

/* vim: set expandtab tabstop=4 shiftwidth=4 softtabstop=4 */

/**
 * CMS Library
 * 
 * PHP version 5.4+
 * 
 * LICENSE: Licensed by AT&T under the 'Software Development Kit Tools 
 * Agreement.' 2013. 
 * TERMS AND CONDITIONS FOR USE, REPRODUCTION, AND DISTRIBUTIONS:
 * http://developer.att.com/sdk_agreement/
 *
 * Copyright 2013 AT&T Intellectual Property. All rights reserved.
 * For more information contact developer.support@att.com
 * 
 * @category  API
 * @package   CMS 
 * @author    pk9069
 * @copyright 2013 AT&T Intellectual Property
 * @license   http://developer.att.com/sdk_agreement AT&amp;T License
 * @link      http://developer.att.com
 */

require_once __DIR__ . '../../Srvc/APIService.php';
require_once __DIR__ . '/CSResponse.php';
require_once __DIR__ . '/SSResponse.php';

use Att\Api\Srvc\Service;
use Att\Api\Srvc\APIService;
use Att\Api\OAuth\OAuthToken;
use Att\Api\Restful\RestfulRequest;
use Att\Api\Restful\HttpPost;

/**
 * Used to interact with version 1 of the CMS API.
 *
 * @category API
 * @package  CMS
 * @author   pk9069
 * @license  http://developer.att.com/sdk_agreement AT&amp;T License
 * @version  Release: @package_version@ 
 * @link     https://developer.att.com/docs/apis/rest/1/Call%20Management%20(Beta)
 */
class CMSService extends APIService
{

    /**
     * Creates a CMSRequest object with the specified URL and access token.
     * 
     * @param string     $FQDN  fully qualified domain name to send requests to
     * @param OAuthToken $token access token to use for authorization
     */
    public function __construct($FQDN, OAuthToken $token)
    {
        parent::__construct($FQDN, $token);
    }

    /**
     * Sends an API request for creating a CMS session. 
     *
     * @param array $vals an array of strings to be send in the request
     *
     * @return CSResponse API response 
     * @throws ServiceException if API request was not successful
     */
    public function createSession($vals) 
    {
        $jvals = json_encode($vals);

        $endpoint = $this->getFqdn() . '/rest/1/Sessions';
        $req = new RestfulRequest($endpoint);
        
        $req
            ->setAuthorizationHeader($this->getToken())
            ->setHeader('Accept', 'application/json')
            ->setHeader('Content-Type', 'application/json');

        $httpPost = new HttpPost();
        $httpPost->setBody($jvals);

        $result = $req->sendHttpPost($httpPost);

        $arr = Service::parseJson($result);
        return CSResponse::fromArray($arr);
    }
    
    /**
     * Sends an API request for sending a signal to a previously created
     * session.
     *
     * @param string $signal    signal to send to a CMS session
     * @param string $sessionId session to which signal will be sent 
     *
     * @return SSResponse API response
     * @throws ServiceException if API request was not successful
     */
    public function sendSignal($signal, $sessionId) 
    {
        $jvals = json_encode(array('signal' => $signal));
        $fqdn = $this->getFqdn();
        $endpoint = $fqdn . '/rest/1/Sessions/' . $sessionId . '/Signals';
        $req = new RESTFulRequest($endpoint);

        $req
            ->setAuthorizationHeader($this->getToken())
            ->setHeader('Accept', 'application/json')
            ->setHeader('Content-Type', 'application/json');

        $httpPost = new HttpPost();
        $httpPost->setBody($jvals);
        $result = $req->sendHttpPost($httpPost);

        $arr = Service::parseJson($result);
        return SSResponse::fromArray($arr);
    }
}

?>
